src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -0,0 +1,624 @@
+    {
+        return std::make_shared<ExtractColumnsTransform>(header, read_from_format_info.requested_columns);
+    });
+
+    auto pipeline = std::make_unique<QueryPipeline>(QueryPipelineBuilder::getPipeline(std::move(builder)));
+    auto current_reader = std::make_unique<PullingPipelineExecutor>(*pipeline);
+
+    ProfileEvents::increment(ProfileEvents::EngineFileLikeReadFiles);
+
+    return ReaderHolder(
+        object_info, std::move(read_buf), std::move(source), std::move(pipeline), std::move(current_reader));
+}
+
+std::future<StorageObjectStorageSource::ReaderHolder> StorageObjectStorageSource::createReaderAsync(size_t processor)
+{
+    return create_reader_scheduler([=, this] { return createReader(processor); }, Priority{});
+}
+
+std::unique_ptr<ReadBuffer> StorageObjectStorageSource::createReadBuffer(const String & key, size_t object_size)
+{
+    auto read_settings = getContext()->getReadSettings().adjustBufferSize(object_size);
+    read_settings.enable_filesystem_cache = false;
+    /// FIXME: Changing this setting to default value breaks something around parquet reading
+    read_settings.remote_read_min_bytes_for_seek = read_settings.remote_fs_buffer_size;
+
+    const bool object_too_small = object_size <= 2 * getContext()->getSettings().max_download_buffer_size;
+    const bool use_prefetch = object_too_small && read_settings.remote_fs_method == RemoteFSReadMethod::threadpool;
+    read_settings.remote_fs_method = use_prefetch ? RemoteFSReadMethod::threadpool : RemoteFSReadMethod::read;
+    /// User's object may change, don't cache it.
+    read_settings.use_page_cache_for_disks_without_file_cache = false;
+
+    // Create a read buffer that will prefetch the first ~1 MB of the file.
+    // When reading lots of tiny files, this prefetching almost doubles the throughput.
+    // For bigger files, parallel reading is more useful.
+    if (use_prefetch)
+    {
+        LOG_TRACE(log, "Downloading object of size {} with initial prefetch", object_size);
+
+        auto async_reader = object_storage->readObjects(
+            StoredObjects{StoredObject{key, /* local_path */ "", object_size}}, read_settings);
+
+        async_reader->setReadUntilEnd();
+        if (read_settings.remote_fs_prefetch)
+            async_reader->prefetch(DEFAULT_PREFETCH_PRIORITY);
+
+        return async_reader;
+    }
+    else
+    {
+        /// FIXME: this is inconsistent that readObject always reads synchronously ignoring read_method setting.
+        return object_storage->readObject(StoredObject(key), read_settings);
+    }
+}
+
+StorageObjectStorageSource::IIterator::IIterator(const std::string & logger_name_)
+    : logger(getLogger(logger_name_))
+{
+}
+
+StorageObjectStorage::ObjectInfoPtr StorageObjectStorageSource::IIterator::next(size_t processor)
+{
+    auto object_info = nextImpl(processor);
+
+    if (object_info)
+    {
+        LOG_TEST(logger, "Next key: {}", object_info->relative_path);
+    }
```

Comment:
Do we want to keep this log?

Reply:
I'd say yes. It was quite useful to me, but removing it will mean `logger` is unused, so removing `logger` will mean we should also remove `logger_name` from ctor, so it becomes not quite quick to return it back (it was useful to me to also see logger name which tells which iterator type is being used for listing keys).